pkg/executor/sortexec/topn.go

```diff
@@ -166,76 +253,319 @@ func (e *TopNExec) loadChunksUntilTotalLimit(ctx context.Context) error {
+		})
 	}
-	newRowChunks.GetMemTracker().SetLabel(memory.LabelForRowChunks)
-	e.memTracker.ReplaceChild(chkHeap.rowChunks.GetMemTracker(), newRowChunks.GetMemTracker())
-	chkHeap.rowChunks = newRowChunks
 
-	e.memTracker.Consume(int64(8 * (len(newRowPtrs) - len(chkHeap.rowPtrs))))
-	chkHeap.rowPtrs = newRowPtrs
+	fetcherWaiter.Wait()
+	workersWaiter.Wait()
 	return nil
 }
+
+func (e *TopNExec) executeTopN(ctx context.Context) {
+	defer func() {
+		if r := recover(); r != nil {
+			processPanicAndLog(e.resultChannel, r)
+		}
+
+		close(e.resultChannel)
+	}()
+
+	heap.Init(e.chkHeap)
+	for uint64(len(e.chkHeap.rowPtrs)) > e.chkHeap.totalLimit {
+		// The number of rows we loaded may exceeds total limit, remove greatest rows by Pop.
+		heap.Pop(e.chkHeap)
+		e.chkHeap.droppedRowNum++
+	}
+
+	if err := e.executeTopNWhenNoSpillTriggered(ctx); err != nil {
+		e.resultChannel <- rowWithError{err: err}
+		return
+	}
+
+	if e.spillHelper.isSpillNeeded() {
+		if err := e.executeTopNWhenSpillTriggered(ctx); err != nil {
+			e.resultChannel <- rowWithError{err: err}
+			return
+		}
+	}
+
+	e.generateTopNResults()
+}
+
+// Return true when spill is triggered
+func (e *TopNExec) generateTopNResultsWHenNoSpillTriggered() bool {
+	rowPtrNum := len(e.chkHeap.rowPtrs)
+	for ; e.chkHeap.idx < rowPtrNum; e.chkHeap.idx++ {
+		if e.chkHeap.idx%10 == 0 && e.spillHelper.isSpillNeeded() {
+			return true
+		}
+		e.resultChannel <- rowWithError{row: e.chkHeap.rowChunks.GetRow(e.chkHeap.rowPtrs[e.chkHeap.idx])}
+	}
+	return false
+}
+
+func (e *TopNExec) generateResultWithMultiWayMerge(offset int64, limit int64) error {
+	multiWayMerge := newMultiWayMerger(&diskSource{sortedRowsInDisk: e.spillHelper.sortedRowsInDisk}, e.lessRow)
+
+	err := multiWayMerge.init()
+	if err != nil {
+		return err
+	}
+
+	outputRowNum := int64(0)
+	for {
+		if limit != -1 && outputRowNum >= limit {
+			return nil
+		}
+
+		row, err := multiWayMerge.next()
+		if err != nil {
+			return err
+		}
+
+		if row.IsEmpty() {
+			return nil
+		}
+
+		if outputRowNum >= offset {
+			select {
+			case <-e.finishCh:
+				return nil
+			case e.resultChannel <- rowWithError{row: row}:
+			}
+		}
+		outputRowNum++
+		injectParallelSortRandomFail(1)
+	}
+}
+
+func (e *TopNExec) generateTopNResultsWhenSpillTriggered() error {
+	inDiskNum := len(e.spillHelper.sortedRowsInDisk)
+	if inDiskNum == 0 {
+		panic("inDiskNum can't be 0 when we generate result with spill triggered")
+	}
+
+	if inDiskNum == 1 {
+		inDisk := e.spillHelper.sortedRowsInDisk[0]
+		chunkNum := inDisk.NumChunks()
+		offset := e.Limit.Offset
+		for i := 0; i < chunkNum; i++ {
+			chk, err := inDisk.GetChunk(i)
+			if err != nil {
+				return err
+			}
+
+			injectTopNRandomFail(10)
+
+			rowNum := chk.NumRows()
+			j := 0
+			if !e.inMemoryThenSpillFlag {
+				// When e.inMemoryThenSpillFlag == false, we need to set j = offset
+				// because rows that should be ignored before offset are also spilled into disk.
+				j = int(offset)
```

Comment:
if `chunkNum > 1`, for every chunk, `j` will always be set to `offset`?

Reply:
> if `chunkNum > 1`, for every chunk, `j` will always be set to `offset`?

I will fix it.